### PR TITLE
Support __{u,}int128_t in radix sort

### PR DESCRIPTION
**Pull request by senior-zero**
This PR supports `__int128_t` and `__uint128_t` in block/device radix sort as well as block radix rank. It's partially answering the [question](https://github.com/NVIDIA/cub/issues/52) on 128 bit types support. Performance wise, `__int128_t` utilize BW similarly to 64 bit types.

### CODE REVIEW

**Review by miscco — APPROVED** [review]
That is awesome, thanks a lot. I assume that `int128` implicitly requires `uint128` which has the same switch?

**Review by elstehle — APPROVED** [review]
Great work, looks solid!

File: test/test_device_radix_sort.cu
Comment (canonizer) [inline test/test_device_radix_sort.cu:908]: Can `base` be of type `UnsignedBits` corresponding to `KeyT`?

File: test/test_util.h
Comment (canonizer) [inline test/test_util.h:732]: Does this have to be defined manually? If yes, there should be tests for it.

File: test/test_util.h
Comment (canonizer) [inline test/test_util.h:734]: I'd call it `MAX_DIGITS`, but this is up to you.

File: test/test_util.h
Comment (canonizer) [inline test/test_util.h:754]: Does this have to be defined manually? If yes, there should be tests for it.

File: test/test_util.h
Comment (canonizer) [inline test/test_util.h:747]: I think that at the beginning of this loop, `digit` points to a position where no digit has been written.

File: test/test_util.h
Comment (canonizer) [inline test/test_util.h:732]: Also, consider having a separate pull request for these two functions, together with the tests.

**Review by canonizer — CHANGES_REQUESTED** [review]

File: test/test_util.h
Comment (senior-zero) [reply test/test_util.h:754]: It's a test util only, therefore tests are welcome but not necessery.

File: test/test_util.h
Comment (senior-zero) [reply test/test_util.h:732]: It's a test util only, therefore tests are welcome but not necessery.

File: cub/util_type.cuh
Comment (senior-zero) [reply cub/util_type.cuh:1208]: This can't be implemented by inheriting from `BaseTraits` because there's no `BaseTraits` specialization for non-primitive types.

File: cub/util_type.cuh
Comment (senior-zero) [reply cub/util_type.cuh:1242]: Same as above

File: test/test_device_radix_sort.cu
Comment (senior-zero) [reply test/test_device_radix_sort.cu]: That's much better, thanks!

File: test/test_util.h
Comment (senior-zero) [reply test/test_util.h:747]: Nice catch, thanks!

File: test/catch2_test_printing.cu
Comment (canonizer) [inline test/catch2_test_printing.cu:1]: Thanks for adding these tests!

File: test/catch2_test_printing.cu
Comment (canonizer) [inline test/catch2_test_printing.cu:17]: Could you also print some strictly positive numbers, and some numbers that don't fit into 64 bits?

File: test/test_device_radix_sort.cu
Comment (canonizer) [inline test/test_device_radix_sort.cu:911]: Will this work if `num_bits == sizeof(KeyT) * 8` and `begin_bit == 0`, e.g. 32 and 0 respectively for `int`?

File: test/test_block_radix_rank.cu
Comment (canonizer) [inline test/test_block_radix_rank.cu:162]: Not sure if this is related to this pull request, but I'll leave it up to you.

File: test/test_device_batch_memcpy.cu
Comment (canonizer) [inline test/test_device_batch_memcpy.cu:316]: Not sure if this is related to this pull request, but I'll leave it up to you.

File: cub/util_type.cuh
Comment (canonizer) [inline cub/util_type.cuh:1217]: Is there any difference from `BaseTraits` apart from `PRIMITIVE == false`? What does `PRIMITIVE` mean and what is it used for?

File: cub/util_type.cuh
Comment (canonizer) [inline cub/util_type.cuh:1252]: Is there any difference from `BaseTraits` apart from `PRIMITIVE == false`? What does `PRIMITIVE` mean and what is it used for?

**Review by canonizer — APPROVED** [review]
Approved, provided that the comments are addressed.

File: test/test_device_radix_sort.cu
Comment (senior-zero) [reply test/test_device_radix_sort.cu:911]: It won't, but we never get this case because this code is under the following condition:
```cpp
if (num_bits < static_cast<int>(sizeof(KeyT) * 8))
```

File: cub/util_type.cuh
Comment (senior-zero) [reply cub/util_type.cuh:1217]: `BaseTraits` has many specializations. If the question is related to differences between `BaseTraits<{UN,}SIGNED_INTEGER, true, false, ...>` specializations and `__{u,}int128_t` specializations, there's no difference apart from `PRIMITIVE == false`.

In my understanding, `PRIMITIVE == true` means that the type is fundamental and not part of CUDA extension. For instance, all vector types are marked as not primitive. Since `__{u,}int128_t`is an extension, it falls into the same category. `PRIMITIVE` is also used in the decoupled look-back as a way to distinguish between types that can be combined with a flag to form a single architectural word or not. So, marking `__{u,}int128_t` as primitive would break scan at the moment.

**Pull request merged**
